presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java

```diff
@@ -0,0 +1,206 @@
+        private final Page page;
+
+        private SelectedPositions selectedPositions;
+        private final Block[] previouslyComputedResults;
+
+        public PositionsPageProcessorIterator(ConnectorSession session, Page page, SelectedPositions selectedPositions)
+        {
+            this.session = session;
+            this.page = page;
+            this.selectedPositions = selectedPositions;
+            this.previouslyComputedResults = new Block[projections.size()];
+        }
+
+        @Override
+        protected Page computeNext()
+        {
+            while (true) {
+                if (selectedPositions.isEmpty()) {
+                    return endOfData();
+                }
+
+                int batchSize = Math.min(selectedPositions.size(), projectBatchSize);
+                Optional<Page> result = processBatch(batchSize);
+
+                // if the page buffer filled up, so halve the batch size and retry
+                if (!result.isPresent()) {
+                    verify(batchSize > 1);
+                    projectBatchSize = projectBatchSize / 2;
+                    continue;
+                }
+
+                Page page = result.get();
+
+                // if we produced a large page, halve the batch size for the next call
+                long pageSize = page.getSizeInBytes();
+                if (page.getPositionCount() > 1 && pageSize > MAX_PAGE_SIZE_IN_BYTES) {
+                    projectBatchSize = projectBatchSize / 2;
+                }
+
+                // if we produced a small page, double the batch size for the next call
+                if (pageSize < MIN_PAGE_SIZE_IN_BYTES && projectBatchSize < MAX_BATCH_SIZE) {
+                    projectBatchSize = projectBatchSize * 2;
+                }
+
+                // remove batch from selectedPositions and previouslyComputedResults
+                selectedPositions = selectedPositions.subRange(batchSize, selectedPositions.size());
+                for (int i = 0; i < previouslyComputedResults.length; i++) {
+                    if (previouslyComputedResults[i] != null && previouslyComputedResults[i].getPositionCount() > batchSize) {
+                        previouslyComputedResults[i] = previouslyComputedResults[i].getRegion(batchSize, previouslyComputedResults[i].getPositionCount() - batchSize);
+                    }
+                    else {
+                        previouslyComputedResults[i] = null;
+                    }
+                }
+
+                return page;
+            }
+        }
+
+        private Optional<Page> processBatch(int batchSize)
+        {
+            Block[] blocks = new Block[projections.size()];
+
+            int pageSize = 0;
+            SelectedPositions positionsBatch = selectedPositions.subRange(0, batchSize);
+            for (int i = 0; i < projections.size(); i++) {
+                if (positionsBatch.size() > 1 && pageSize > MAX_PAGE_SIZE_IN_BYTES) {
+                    return Optional.empty();
```

Comment:
@dain Why It is not OK to produce few too large pages until `projectBatchSize` stabilizes?
I am thinking it is possible to have data that will cause `https://github.com/prestodb/presto/pull/7515/files/5ca5f9440235ef3822961e7ad8e273b32c1de522#diff-92486801dfdc982c80726599c9c31357R179` to recompute results for every second input page, which is probably not desirable.